File: src/main/java/io/github/shaksternano/mediamanipulator/image/io/reader/JavaxImageReader.java
package io.github.shaksternano.mediamanipulator.image.io.reader;

import com.google.common.collect.ImmutableSet;
import io.github.shaksternano.mediamanipulator.image.imagemedia.ImageMedia;
import io.github.shaksternano.mediamanipulator.image.imagemedia.StaticImage;
import io.github.shaksternano.mediamanipulator.image.util.ImageUtil;
import org.jetbrains.annotations.Nullable;

import javax.imageio.ImageIO;
import java.awt.image.BufferedImage;
import java.io.IOException;
import java.io.InputStream;
import java.util.Set;

public class JavaxImageReader implements ImageReader {

    @Override
    public ImageMedia read(InputStream inputStream, @Nullable Integer type) throws IOException {
        BufferedImage image = ImageIO.read(inputStream);

        if (image == null) {
            throw new IOException();
        } else {
            if (type != null) {
                image = ImageUtil.convertType(image, type);
            }

            return new StaticImage(image);
        }
    }

    @Override
    public Set<String> getSupportedFormats() {
        return ImmutableSet.of(
                "bmp",
                "jpeg",
                "jpg",
                "wbmp",
                "png",
                "gif",
                "tif",
                "tiff"
        );
    }
}

File: core/src/main/kotlin/media/reader/JavaxImageReader.kt
package com.shakster.borgar.core.media.reader

import com.shakster.borgar.core.io.DataSource
import com.shakster.borgar.core.io.IO_DISPATCHER
import com.shakster.borgar.core.media.ImageFrame
import com.shakster.borgar.core.media.ImageReaderFactory
import com.shakster.borgar.core.media.addTransparency
import kotlinx.coroutines.flow.Flow
import kotlinx.coroutines.flow.flowOf
import kotlinx.coroutines.withContext
import java.awt.image.BufferedImage
import javax.imageio.ImageIO
import kotlin.time.Duration
import kotlin.time.Duration.Companion.milliseconds

class JavaxImageReader(
    image: BufferedImage,
) : BaseImageReader() {

    private val frame: ImageFrame = run {
        // For some reason some images have a greyscale type, even though they have color
        val converted = image.addTransparency()
        ImageFrame(converted, 1.milliseconds, Duration.ZERO)
    }

    override val frameCount: Int = 1
    override val frameRate: Double = 1.0
    override val duration: Duration = 1.milliseconds
    override val frameDuration: Duration = 1.milliseconds
    override val width: Int = frame.content.width
    override val height: Int = frame.content.height
    override val loopCount: Int = 0

    override suspend fun readFrame(timestamp: Duration): ImageFrame = frame

    override fun asFlow(): Flow<ImageFrame> = flowOf(frame)

    override suspend fun reversed(): MediaReader<ImageFrame> = this

    override suspend fun createChangedSpeed(speedMultiplier: Double): MediaReader<ImageFrame> = this

    override suspend fun close() = Unit

    object Factory : ImageReaderFactory {
        override val supportedFormats: Set<String> = setOf(
            "bmp",
            "jpeg",
            "jpg",
            "wbmp",
            "png",
            "gif",
            "tif",
            "tiff",
        )

        override suspend fun create(input: DataSource): ImageReader = input.newStream().use {
            val image = withContext(IO_DISPATCHER) {
                ImageIO.read(it)
            } ?: throw IllegalArgumentException("Failed to read image")
            JavaxImageReader(image)
        }
    }
}
